File: src/offerings/offering.py
import uuid
from typing import Optional, List

from src.Users.instructor import Instructor


class Offering:
    def __init__(self, location, schedule, lesson_type):
        self.__isAvailable = False
        self.__location = location
        self.__schedule = schedule
        self.__lesson_type = lesson_type
        self.__instructor = None
        self.__id = generate_unique_offering_id()

    def __eq__(self, other):
        if isinstance(other, Offering):
            return (self.__location, self.__schedule) == (
                other.__location,
                other.__schedule,
            )
        return False

    def get_id(self) -> str:
        return self.__id

    def check_available(self):
        return self.__isAvailable

    def assign_instructor(self, instructor):
        if self.check_available():
            self.__instructor = instructor
            self.__isAvailable = False
            return True
        return False

    def get_location(self):
        return self.__location

    def get_schedule(self):
        return self.__schedule

    def get_instructor(self) -> Optional[Instructor]:
        return self.__instructor

    def __hash__(self):
        return hash(self.__id)


class OfferingCatalog:
    __offerings: List[Offering] = {}

    @staticmethod
    def is_offering_unique(location, schedule):
        for offering in OfferingCatalog.__offerings:
            if (
                offering.get_location() == location
                and offering.get_schedule() == schedule
            ):
                return False
        return True

    @staticmethod
    def create_offering(location, schedule, lesson_type):
        if OfferingCatalog.is_offering_unique(location, schedule):
            new_offering = Offering(location, schedule, lesson_type)
            OfferingCatalog.add(new_offering)
            return new_offering
        return None

    @staticmethod
    def add(offering):
        OfferingCatalog.__offerings.append(offering)

    @staticmethod
    def get_all_unassigned_offerings():
        return [
            offering
            for offering in OfferingCatalog.__offerings
            if offering.get_instructor() is None
        ]

    @staticmethod
    def select_offering(offering_id: str, instructor: Instructor) -> Optional[Offering]:
        if offering_id in OfferingCatalog.__offerings:
            offering = OfferingCatalog.__offerings[offering_id]
            if offering.assign_instructor(instructor):
                return offering
        return None


def generate_unique_offering_id():
    return str(uuid.uuid4())

File: src/main/java/offerings/Offering.java
package offerings;

import users.Client;
import users.Instructor;

import java.util.*;

public class Offering {

    private boolean isAvailableToPublic;
    private boolean hasInstructor;
    private final Location location;
    private final Schedule schedule;
    private final String lessonType;
    private Instructor instructor;
    private final int id;
    private final List<Client> registeredClients;
    private int max_size = 3; // placeholder value
    private Set<Integer> registeredClientIds;

    public Offering(Location location, Schedule schedule, String lessonType) {
        this(generateUniqueOfferingId(), location, schedule, lessonType, false, null);
    }

    public Offering(int offeringId, Location location, Schedule schedule, String lessonType, boolean isAvailableToPublic, Instructor instructor, int max_size, List<Client> registeredClients, Set<Integer> registeredClientIds) {
        this.registeredClients = new ArrayList<>(registeredClients);
        this.isAvailableToPublic = isAvailableToPublic;
        this.location = location;
        this.schedule = schedule;
        this.lessonType = lessonType;
        this.id = offeringId;
        hasInstructor = instructor != null;
        this.instructor = instructor;
        this.max_size = max_size;
        this.registeredClientIds = new HashSet<>(registeredClientIds);
    }

    public Offering(int offeringId, Location location, Schedule schedule, String lessonType, boolean isAvailableToPublic, Instructor instructor) {
        this.registeredClients = new ArrayList<>();
        this.isAvailableToPublic = isAvailableToPublic;
        this.location = location;
        this.schedule = schedule;
        this.lessonType = lessonType;
        this.id = offeringId;
        hasInstructor = instructor != null;
        this.instructor = instructor;
    }

    public void updateAvailability(boolean availability) {
        this.isAvailableToPublic = availability;
    }

    public boolean hasInstructor() {
        return hasInstructor;
    }

    public int getId() {
        return id;
    }

    // Method to check availability
    public boolean checkAvailable() {
        return isAvailableToPublic;
    }

    public boolean assignInstructor(Instructor instructor) {
        if (!hasInstructor) {
            this.instructor = instructor;
            this.hasInstructor = true;
            this.isAvailableToPublic = true;
            return true;
        }
        return false;
    }

    public boolean registerClient(Client client) {
        if (!registeredClients.contains(client) && registeredClients.size() <= max_size) {
            registeredClients.add(client);
            if (registeredClients.size() >= max_size) {
                isAvailableToPublic = false;
            }
            return true;
        }
        return false;
    }

    // Getter for location
    public Location getLocation() {
        return location;
    }

    // Getter for schedule
    public Schedule getSchedule() {
        return schedule;
    }

    // Getter for instructor
    public Instructor getInstructor() {
        return instructor;
    }

    // Getter for lessonType
    public String getLessonType() {
        return lessonType;
    }

    public boolean isAvailableToPublic() {
        return isAvailableToPublic;
    }

    @Override
    public int hashCode() {
        return Integer.hashCode(id);
    }

    @Override
    public boolean equals(Object other) {
        if (other instanceof Offering) {
            Offering otherOffering = (Offering) other;
            return this.location.equals(otherOffering.location)
                    && this.schedule.equals(otherOffering.schedule);
        }
        return false;
    }

    public String toString() {
        return "Offering for Lesson: " + lessonType + " at Location: " + location + ", at time: " + schedule;
    }

    // Generate a unique offering ID
    private static int generateUniqueOfferingId() {
        return UUID.randomUUID().toString().hashCode();
    }

    public int getMaxSize() {
        return max_size;
    }

    public Set<Integer> getClientIdSet() {
        return registeredClientIds;
    }

    public List<Client> getClients() {
        return registeredClients;
    }
}
